pkg/controller/batchrelease/labelpatch/patcher.go

```diff
@@ -55,59 +57,85 @@ func (r *realPatcher) PatchPodBatchLabel(ctx *batchcontext.BatchContext) error {
 
 // PatchPodBatchLabel will patch rollout-id && batch-id to pods
 func (r *realPatcher) patchPodBatchLabel(pods []*corev1.Pod, ctx *batchcontext.BatchContext) error {
-	// the number of active pods that has been patched successfully.
-	patchedUpdatedReplicas := int32(0)
-	// the number of target active pods that should be patched batch label.
-	plannedUpdatedReplicas := ctx.PlannedUpdatedReplicas
+	plannedUpdatedReplicasForBatches := r.calculatePlannedStepIncrements(ctx.Batches, int(ctx.Replicas), int(ctx.CurrentBatch))
+	var updatedButUnpatchedPods []*corev1.Pod
 
 	for _, pod := range pods {
+		if !pod.DeletionTimestamp.IsZero() {
+			klog.InfoS("Pod is being deleted, skip patching", "pod", klog.KObj(pod), "rollout", r.logKey)
+			continue
+		}
+		// we don't patch label for the active old revision pod
 		if !util.IsConsistentWithRevision(pod, ctx.UpdateRevision) {
+			klog.InfoS("Pod is not consistent with revision, skip patching", "pod", klog.KObj(pod), "rollout", r.logKey)
 			continue
 		}
-
-		podRolloutID := pod.Labels[v1beta1.RolloutIDLabel]
-		if pod.DeletionTimestamp.IsZero() && podRolloutID == ctx.RolloutID {
-			patchedUpdatedReplicas++
+		if pod.Labels[v1beta1.RolloutIDLabel] != ctx.RolloutID {
+			// for example: new/recreated pods
+			updatedButUnpatchedPods = append(updatedButUnpatchedPods, pod)
+			klog.InfoS("Find a pod to add updatedButUnpatchedPods", "pod", klog.KObj(pod), "rollout", r.logKey)
+			continue
 		}
-	}
 
-	// all pods that should be patched have been patched
-	if patchedUpdatedReplicas >= plannedUpdatedReplicas {
-		return nil // return fast
+		podBatchID, err := strconv.Atoi(pod.Labels[v1beta1.RolloutBatchIDLabel])
+		if err != nil {
+			klog.InfoS("Pod batchID is not a number, skip patching", "pod", klog.KObj(pod), "rollout", r.logKey)
+			continue
+		}
+		plannedUpdatedReplicasForBatches[podBatchID-1]--
 	}
-
-	for _, pod := range pods {
-		if pod.DeletionTimestamp.IsZero() {
-			// we don't patch label for the active old revision pod
-			if !util.IsConsistentWithRevision(pod, ctx.UpdateRevision) {
-				continue
+	klog.InfoS("updatedButUnpatchedPods amount calculated", "amount", len(updatedButUnpatchedPods), "rollout", r.logKey)
+	// patch the pods
+	for i := len(plannedUpdatedReplicasForBatches) - 1; i >= 0; i-- {
+		for ; plannedUpdatedReplicasForBatches[i] > 0; plannedUpdatedReplicasForBatches[i]-- {
+			if len(updatedButUnpatchedPods) == 0 {
+				return fmt.Errorf("no pods to patch for %v, batch %d", r.logKey, i+1)
```

Comment:
no need to report error if no more pods to patch, the workload may be scaled down during the rollout

Reply:
done